sw/device/lib/aes.c

```diff
@@ -0,0 +1,100 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/aes.h"
+
+#include "aes_regs.h"  // Generated.
+#include "sw/device/lib/common.h"
+
+#define AES0_BASE_ADDR 0x40110000
+
+void aes_init(aes_cfg_t aes_cfg) {
+  // Convert key_len from bytes to onehot
+  size_t key_len_onehot = 0b001;
+  if (aes_cfg.key_len_in_bytes == 24) {
+    key_len_onehot = 0b010;
+  } else if (aes_cfg.key_len_in_bytes == 32) {
+    key_len_onehot = 0b100;
+  }
+
+  REG32(AES_CTRL(0)) =
+      aes_cfg.mode << AES_CTRL_MODE |
+      (key_len_onehot & AES_CTRL_KEY_LEN_MASK) << AES_CTRL_KEY_LEN_OFFSET |
+      aes_cfg.manual_start_trigger << AES_CTRL_MANUAL_START_TRIGGER |
+      aes_cfg.force_data_overwrite << AES_CTRL_FORCE_DATA_OVERWRITE;
+};
+
+void aes_key_put(const void *key, size_t key_len_in_bytes) {
+  const size_t key_len_in_words = key_len_in_bytes >> 2;
+
+  for (int i = 0; i < key_len_in_words; ++i) {
+    REG32(AES_KEY0(0) + i * sizeof(uint32_t)) = ((uint32_t *)key)[i];
+  }
+  for (int i = key_len_in_words; i < 8; ++i) {
+    REG32(AES_KEY0(0) + i * sizeof(uint32_t)) = 0x0;
+  }
+}
+
+void aes_data_put_wait(const void *data) {
+  // Wait for AES unit to be ready for new input data
+  aes_data_ready();
+
+  // Provide the input data
+  aes_data_put(data);
+}
+
+void aes_data_put(const void *data) {
+  for (int i = 0; i < 4; ++i) {
+    REG32(AES_DATA_IN0(0) + i * sizeof(uint32_t)) = ((uint32_t *)data)[i];
+  }
+}
+
+void aes_data_get_wait(void *data) {
+  // Wait for AES unit to have valid output data
+  aes_data_valid();
+
+  // Get the data
+  aes_data_get(data);
+}
+
+void aes_data_get(void *data) {
+  for (int i = 0; i < 4; ++i) {
+    ((uint32_t *)data)[i] = REG32(AES_DATA_OUT0(0) + i * sizeof(uint32_t));
+  }
+}
+
+void aes_data_ready(void) {
+  while (!((REG32(AES_STATUS(0)) >> AES_STATUS_INPUT_READY) & 0x1)) {
+  }
+}
+
+void aes_data_valid(void) {
+  while (!((REG32(AES_STATUS(0)) >> AES_STATUS_OUTPUT_VALID) & 0x1)) {
+  }
+}
+
+void aes_idle(void) {
+  while (!((REG32(AES_STATUS(0)) >> AES_STATUS_IDLE) & 0x1)) {
+  }
+}
+
+void aes_clear(void) {
+  // Wait for idle
+  aes_idle();
+
+  // Disable autostart
+  REG32(AES_CTRL(0)) = 0x1 << AES_CTRL_MANUAL_START_TRIGGER;
+
+  // Clear internal key and output registers
+  REG32(AES_TRIGGER(0)) = (0x1 << AES_TRIGGER_KEY_CLEAR) |
+                          (0x1 << AES_TRIGGER_DATA_IN_CLEAR) |
+                          (0x1 << AES_TRIGGER_DATA_OUT_CLEAR);
+
+  // Wait for output not valid, and input ready
+  uint32_t status = (0x1 << AES_STATUS_OUTPUT_VALID);
+  while (((status >> AES_STATUS_OUTPUT_VALID) & 0x1) ||
```

Comment:
Consider splitting out "((status >> AES_STATUS_OUTPUT_VALID) & 0x1)" in its own function, similar like aes_data_valid. And also "!((status >> AES_STATUS_INPUT_READY) & 0x1)"

Reply:
Yep, thanks to implementing your previous suggestions, this one becomes very easy. I changed it.